src/Integration/Aws/SimpleS3/src/SimpleS3Client.php

```diff
@@ -21,8 +26,64 @@ public function getUrl(string $bucket, string $key): string
         return $this->getEndpoint($uri, [], null);
     }
 
-    /*
-     * TODO Add get() and put() which supports large files
-     * TODO add has()
+    public function download(string $bucket, string $key): ResultStream
+    {
+        return $this->getObject(['Bucket' => $bucket, 'Key' => $key])->getBody();
+    }
+
+    public function has(string $bucket, string $key): bool
+    {
+        return $this->objectExists(['Bucket' => $bucket, 'Key' => $key])->isSuccess();
+    }
+
+    /**
+     * @param string|resource|callable|iterable $object
      */
+    public function upload(string $bucket, string $key, $object): void
+    {
+        $chunkSize = 64 * 1024;
+        $directUpload = \is_string($object);
+        if (!$directUpload && $isResource = \is_resource($object)) {
+            // Try to find size of the resource
+            $stats = \fstat($object);
+            // If we dont know the size, or if the size is larger than chunk size
+            if (!isset($stats['size']) || $stats['size'] > $chunkSize) {
+                $directUpload = false;
+            }
+        }
+
+        if ($directUpload) {
+            $this->putObject(['Bucket' => $bucket, 'Key' => $key, 'Body' => $object]);
+
+            return;
+        }
+
+        if ($isResource) {
+            $stream = ResourceStream::create($object, $chunkSize);
+        } elseif (\is_callable($object)) {
+            $stream = CallableStream::create($object, $chunkSize);
+        } elseif (\is_iterable($object)) {
+            $stream = IterableStream::create($object);
```

Comment:
Is it okey to use the streams here?

They are really helpful =)

Reply:
Yeah they are really helpful in that context !